src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSSynchronizationContext.cs

```diff
@@ -0,0 +1,140 @@
+
+namespace System.Runtime.InteropServices.JavaScript {
+    internal sealed unsafe class JSSynchronizationContext : SynchronizationContext {
+        public readonly Thread MainThread;
+
+        internal readonly struct WorkItem {
+            public readonly SendOrPostCallback Callback;
+            public readonly object? Data;
+            public readonly ManualResetEventSlim? Signal;
+
+            public WorkItem (SendOrPostCallback callback, object? data, ManualResetEventSlim? signal) {
+                Callback = callback;
+                Data = data;
+                Signal = signal;
+            }
+        }
+
+        private static JSSynchronizationContext? MainThreadSynchronizationContext;
+        private readonly QueueType Queue;
+        private readonly Action _DataIsAvailable;
+
+        private JSSynchronizationContext (Thread mainThread)
+            : this (
+                mainThread,
+                Channel.CreateUnbounded<WorkItem>(
+                    new UnboundedChannelOptions { SingleWriter = false, SingleReader = true, AllowSynchronousContinuations = true }
+                )
+            )
+        {
+        }
+
+        private JSSynchronizationContext (Thread mainThread, QueueType queue) {
+            MainThread = mainThread;
+            Queue = queue;
+            _DataIsAvailable = DataIsAvailable;
+        }
+
+        public override SynchronizationContext CreateCopy () {
+            return new JSSynchronizationContext(MainThread, Queue);
+        }
+
+        private void AwaitNewData () {
+            var vt = Queue.Reader.WaitToReadAsync();
+            if (vt.IsCompleted) {
+                DataIsAvailable();
+                return;
+            }
+
+            // Once data is added to the queue, vt will be completed on the thread that added the data
+            //  because we created the channel with AllowSynchronousContinuations = true. We want to
+            //  fire a callback that will either schedule a background job to pump (if this is the main
+            //  thread) or use the emscripten API to push a pump request to the main thread.
+            var awaiter = vt.AsTask().ConfigureAwait(false).GetAwaiter();
+            // UnsafeOnCompleted avoids spending time flowing the execution context (we don't need it.)
+            awaiter.UnsafeOnCompleted(_DataIsAvailable);
+        }
+
+        private void DataIsAvailable () {
+            // While we COULD pump here, we don't want to. We want the pump to happen on the next event loop turn.
+            // Otherwise we could get a chain where a pump generates a new work item and that makes us pump again, forever.
+            ScheduleBackgroundJob((void*)(delegate* unmanaged[Cdecl]<void>)&BackgroundJobHandler);
+        }
+
+        public override void Post (SendOrPostCallback d, object? state) {
+            var workItem = new WorkItem(d, state, null);
+            while (!Queue.Writer.TryWrite(workItem))
```

Comment:
When would this ever return false? The channel is unbounded.

Reply:
The documentation was unclear on when this could fail, and I didn't have a chance to read it yet since there are multiple queue implementations. Wasn't sure if it was a wait-free thing where TryWrite failing means 'try again, someone had the lock'

Reply:
cc @IEvangelist a chance to improve the docs! Though this might be an API doc.